### PR TITLE
Add dropdown module

### PR DESCRIPTION
**Pull request by domenkozar**
Add dropdown module
    
    Most of code I've gotten from @ocharles, I've made some minor
    corrections to documentation (typos, wrong examples).

Depends on hercules-ci/elm-dropdown, which is just released version of circuithub/elm-dropdown.

### CODE REVIEW

**Comment by ocharles** [comment]
I will try and port all our code over to this branch to double check. The API might need further refinements, but I think we can still merge this PR if our code also compiles.

**Comment by domenkozar** [comment]
Pushed a commit that allows overriding toggle icon

File: SemanticUI/Modules/Dropdown.elm [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+Amongst other things it adds the "ui dropdown" class to the root node.
+
+-}
+root : Config r msg -> State -> Render msg -> List (Attribute msg) -> List (Html msg) -> Html msg
+root cfg state render attrs children =
+    let
+        ctx =
+            toContext cfg state
+
+        attrs_ =
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown.elm:259]: How about `finalAttrs`? (Since it's not idiomatic to elm to use prime marks)

File: SemanticUI/Modules/Dropdown.elm [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+  - readOnly - Is the dropdown read only (default False)
+
+-}
+type alias Config r msg =
+    { r
+        | identifier : String
+        , onToggle : State -> msg
+        , toggleEvent : ToggleEvent
+        , readOnly : Bool
+    }
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown.elm:174]: I think these `Config` types probably wants to follow the same un-extended style as the existing `Checkbox` component?

File: SemanticUI/Modules/Dropdown.elm [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+The final resultant value is what you decide (some kind of DOM)
+
+-}
+type alias Layout msg html =
+    { makeDropdown : WrappedNode msg
+    , makeToggle : WrappedNode msg
+    , makeDrawer : WrappedNode msg
+    , makeItem : WrappedNode msg
+    }
+    -> html
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown.elm:149]: I was initially wondering if these want to be called 

```
    { renderDropdown : ...
    , renderToggle : ...
    , renderDrawer : ...
    , renderItem : ...
    }
```

I'm a little bit unsure of the `WrappedNode` type... I think I get a bit worried when I see we need a type alias for a function signature.

It feels somewhat related to `HtmlDetails` in elm-sortable-table... Perhaps worth comparing? I think one difference was that `HtmlDetails` in elm-sortable-table doesn't allow one to specify the dom node

File: SemanticUI/Modules/Dropdown/Select.elm [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+inline : Bool -> { a | inline : Bool } -> { a | inline : Bool }
+inline b a =
+    { a | inline = b }
+
+
+type alias State r a =
+    { r
+        | dropdownState : Dropdown.State
+        , selectedValue : a
+    }
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown/Select.elm:115]: I suppose Checkbox also uses this extended style for State

```elm
checkbox : Config msg -> { r | checked : Bool } -> Html msg
```

...but I personally just end up using `State {} Foo`. E.g.

```elm
dropdown1State : Select.State {} (Maybe String)
dropdown2State : Select.State {} Int
dropdown3State : Select.State {} Status
```

I suppose this style could be somewhat useful if one wants to extend the dropdown module with extra generic functionality. I feel like I probably don't mind if the arguments to the functions are anonymous types in this style, but do we want the exposed types to carry around this `r` parameter?

File: SemanticUI/Modules/Dropdown.elm (hunk 3 above)
Comment (rehno-lindeque) [reply SemanticUI/Modules/Dropdown.elm:149]: ~It's also a little unclear to me what `Layout` means, I feel like the layout is somewhat external to these functions?~

File: SemanticUI/Modules/Dropdown.elm [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+    render (class "item" :: attrs) children
+
+
+type alias Context msg =
+    { drawerIsOpen : Bool
+    , drawerIsTransitioning : Bool
+    , drawerIsVisible : Bool
+    , drawerToggledFinal : State
+    , dropdownCfg : Dropdown.Config msg
+    }
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown.elm:394]: Not sure about the name `Context`, this seems like `State` to me. Should `State` be renamed to `DrawerState` perhaps?

File: SemanticUI/Modules/Dropdown/Select.elm [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+as part of the list.
+
+-}
+select :
+    Config r0 a a msg
+    -> State r1 a
+    -> Render msg
+    -> List (Attribute msg)
+    -> (WrappedNode msg -> Bool -> a -> Html msg)
+    -> Html msg
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown/Select.elm:147]: I feel as though it's a little inconsistent that there's a `Layout` type alias for `dropdown`, but not for `select` or `selectMaybe`. 

Not quite sure what to do about it though...

File: SemanticUI/Modules/Dropdown.elm (hunk 5 above)
Comment (rehno-lindeque) [reply SemanticUI/Modules/Dropdown.elm:394]: Ah right I see now that this is derived state. Still feels a little odd to me that it's mixed with Dropdown.Config here but I see why...

File: SemanticUI/Modules/Dropdown.elm [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+                                    state
+
+                                else
+                                    Opening
+
+                            else if not drawerIsOpen then
+                                state
+
+                            else
+                                Closing
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown.elm:447]: I tend to like listing out the truth table for cases like this where every combination needs to be analyzed. Ollie, would you mind if I changed it to read like this?

```elm
case ( toggleOpen, isOpen ) of
    ( True, False ) ->
        Opening

    ( False, True ) ->
        Closing

    _ ->
        state

```

File: SemanticUI/Modules/Dropdown.elm [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+    }
+    -> html
+
+
+{-| Define when the dropdown should open
+-}
+type ToggleEvent
+    = OnClick
+    | OnHover
+    | OnFocus
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown.elm:157]: This could be 

```elm
type alias ToggleEvent = Dropdown.ToggleEvent
```

...but I suppose this is necessary in order to re-export the contructors? It's a shame that we have to jump through hoops to do that.

File: SemanticUI/Modules/Dropdown.elm (hunk 3 above)
Comment (rehno-lindeque) [reply SemanticUI/Modules/Dropdown.elm:149]: Oops I see now that this does do the layout. I didn't read the type properly the first time around

Comment (rehno-lindeque) [reply SemanticUI/Modules/Dropdown.elm:149]: Nevermind... Now that I've loaded all this in my head again I think this is doing the right thing after all. Might just look into the detail a bit more

Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown.elm:149]: ~Can we somehow skip makeDropdown, since it's generally the root element?~

No, looks like `makeToggle (makeDropdown div) ....` is a typical use

File: SemanticUI/Modules/Common.elm [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+It takes the following:
+
+  - Node type as a render function e.g. `Html.div`
+  - Extra attributes of node
+  - The children of node
+
+-}
+type alias WrappedNode msg =
+    Render msg -> List (Attribute msg) -> List (Html msg) -> Html msg
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Common.elm:22]: I think this should be something like `WrappedElement`, because `Html.text` is also a DOM node, but we're only interested in nodes that take attrs and child nodes as arguments.

Comment (rehno-lindeque) [reply SemanticUI/Modules/Common.elm:22]: Then again elm has `Html.node` and `VirtualDom.node` functions, so I see where this comes from. But I think it's the virtualdom/html packages that are mistaken in this case.

> An element is a specific type of node, one that can be directly specified in the HTML with an HTML tag and can have properties like an id or a class. can have children, etc... There are other types of nodes such as comment nodes, text nodes, etc... with different characteristics. Each node has a property .nodeType which reports what type of node it is.

From https://stackoverflow.com/a/9979779/167485

File: SemanticUI/Modules/Common.elm [hunk 10]
```diff
@@ -0,0 +1,36 @@
+module SemanticUI.Modules.Common exposing (Render, WrappedContent, WrappedNode)
+
+import Html exposing (..)
+
+
+{-| Convenience alias for a HTML DOM Node render function
+-}
+type alias Render msg =
+    List (Attribute msg) -> List (Html msg) -> Html msg
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Common.elm:9]: It feels to me like maybe this might be `Renderer` or something like that, similar to `Builder`/`Parser`/`Decoder` etc... (?)

**Comment by ocharles** [comment]
@rehno-lindeque It feels to me a lot of these comments have nothing to do with @domenkozar's work, so it seems unfair to have him fix them. I suggest we continue with the plan of deleting our code, adopting this code, and then if we have things we want to change in the future, we can do that in a later PR. This is really just the original code @HanStolpo wrote, afaik.

**Comment by rehno-lindeque** [comment]
Sorry yeah, I mentioned in Slack, but should have mentioned here as well. I'm actively working on this myself and trying to solicit feedback / thinking out loud. Sorry for the bikeshedding noise @domenkozar (but feel free to weigh in if you want!).

I'm planning to have a new PR up and drop into our code if everyone is happy with it. This is a really surprisingly complicated little component, I feel as though some of the patterns here are going to repeat for other components.

File: SemanticUI/Modules/Dropdown/Select.elm [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+
+-}
+type alias Config r c s msg =
+    { r
+        | identifier : String
+        , onToggle : Dropdown.State -> msg
+        , toggleEvent : Dropdown.ToggleEvent
+        , readOnly : Bool
+        , onSelect : s -> msg
+        , choices : List c
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown/Select.elm:67]: Would people be happy calling this `options` and `makeOption` rather than `choices` and `makeChoice`?

I believe the traditional html control would look something like

```
<select> 
  <option>A</option>
  <option>B</option>
</select>
```

**Comment by rehno-lindeque** [comment]
After chipping away at the `select` component for a while now, I've come to the conclusion that it's not quite right...

A bit of background first. Basically, semantic ui `selection` works like this:

* SemanticUI javascript will dynamically wrap a `<select class="ui dropdown">` with `<div class="ui selection">`
* It will visually connect the drawer to the root element. The drawer has to be a direct child of the root for this to style correctly.

Our select control appears to have two competing goals:

* It's trying to be a very generic selectable dropdown control. But this is essentially just a dropdown with all the weird quirks and layouts that a regular dropdown has.
* Simultaneously it's  a semantic ui `selection` control. But for our purposes `selection` is really just a visual style in semantic UI.

Basically the "selectionable" aspect of the dropdown doesn't have all that much to do with the specific styling / layout of the control... 

When I was initially trying to decide between `dropdown` and `select` I was rather confused about what exactly the trade off is that I'm making. Almost every dropdown is some kind of selectable thing, so why not just use `select` everywhere? But then sometimes I'll need to drop `select` in favour of `dropdown` because I want to customize something. So maybe just stick with `dropdown`?

All that being said, the select control is doing some fairly useful things I think. I'm just not convinced it should go into a library just yet.

**Comment by HanStolpo** [comment]
@rehno-lindeque the point of the that control and all the stuff is to give you the semantic-ui dropdown module without including their JS. It does the expand animation etc. There are no competing goals here as far as I can remember and it is specific to semantic-ui, in semantic its used to build expandle menus as well and depending on where you use it the elements need to be certain type IIRC, in our code we usually only use the select.

The interface is not in the style of the rest of the stuff in the semantic-ui library, so that is probably more useful to change for consistency.

**Comment by rehno-lindeque** [comment]
@HanStolpo understood about the use-case for the `select` control (especially in our own code).

All I mean is that many possible configurations that are based on selection (e.g. https://semantic-ui.com/modules/dropdown.html#search-in-menu) isn't, as far as I know, supported by the `select` component.

Basically I think that since this is in a library people will ask for new features (like the configurable icon in this PR) that will lead us back to the ultra-configurable format that `dropdown` is in.

File: SemanticUI/Modules/Dropdown/Select.elm [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+        { identifier = cfg.identifier
+        , onToggle = cfg.onToggle
+        , toggleEvent = cfg.toggleEvent
+        , readOnly = cfg.readOnly
+        }
+        st.dropdownState
+        (\{ makeDropdown, makeToggle, makeDrawer, makeItem } ->
+            makeDropdown renderRoot
+                (classList [ ( "selection", not cfg.inline ), ( "inline", cfg.inline ) ] :: rootAttrs)
+                [ makeToggle input [ type_ "hidden" ] []
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown/Select.elm:159]: If I'm not mistaken, I think the reason for this input field is that it will include the value in a form's submit function.

We would need to set the `name` and `value` attributes for this to work.

File: SemanticUI/Modules/Dropdown/Select.elm [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+                , layoutChoice (\r a c -> makeToggle r (classList [ ( "text", cfg.inline ) ] :: a) c) False st.selectedValue
```
Comment (rehno-lindeque) [inline SemanticUI/Modules/Dropdown/Select.elm:160]: It looks like the "text" class is needed for all labeled ("labeled" / "selection" / "inline") dropdowns to correctly apply styling on its children, similart to the "item" class in the drawer.
